src/debug/injected/recorder.ts

```diff
@@ -92,24 +92,69 @@ export class Recorder {
     });
   }
 
-  private _buildSelector(node: Node): string {
-    const element = node as Element;
-    for (const attribute of ['data-testid', 'aria-label', 'id', 'data-test-id', 'data-test']) {
+  private _buildSelector(targetElement: Element): string {
+    const path: string[] = [];
+    const root = document.documentElement;
+    for (let element: Element | null = targetElement; element && element !== root; element = element.parentElement) {
+      const selector = this._buildSelectorCandidate(element);
+      if (selector)
+        path.unshift(selector.selector);
+
+      const fullSelector = path.join(' ');
+      if (selector && selector.final)
+        return fullSelector;
+      if (targetElement === this._injectedScript.querySelector(parseSelector(fullSelector), root))
+        return fullSelector;
+    }
+    return '<selector>';
+  }
+
+  private _buildSelectorCandidate(element: Element): { final: boolean, selector: string } | null {
+    for (const attribute of ['data-testid', 'aria-label', 'data-test-id', 'data-test']) {
       if (element.hasAttribute(attribute))
-        return `[${attribute}=${element.getAttribute(attribute)}]`;
+        return { final: true, selector: `${element.nodeName.toLocaleLowerCase()}[${attribute}=${element.getAttribute(attribute)}]` };
     }
     if (element.nodeName === 'INPUT') {
       if (element.hasAttribute('name'))
-        return `[input name=${element.getAttribute('name')}]`;
+        return { final: false, selector: `input[name=${element.getAttribute('name')}]` };
       if (element.hasAttribute('type'))
-        return `[input type=${element.getAttribute('type')}]`;
+        return { final: false, selector: `input[type=${element.getAttribute('type')}]` };
+    } else if (element.nodeName === 'IMG') {
+      if (element.hasAttribute('alt'))
+        return { final: false, selector: `img[alt="${element.getAttribute('alt')}"]` };
     }
-    if (element.firstChild && element.firstChild === element.lastChild && element.firstChild.nodeType === Node.TEXT_NODE)
-      return `text="${element.textContent}"`;
-    return '<selector>';
+    const textSelector = textSelectorForElement(element);
+    if (textSelector)
+      return { final: false, selector: textSelector };
+
+    // Depreoritize id, but still use it as a last resort.
+    if (element.hasAttribute('id'))
+      return { final: true, selector: `${element.nodeName.toLocaleLowerCase()}[id=${element.getAttribute('id')}]` };
```

Comment:
With shadow dom, id does not guarantee uniqueness anymore. Perhaps, do not mark it as final?

Reply:
Nothing works for shadow yet